File: uri/estructuras/elementary_my_dear_watson/1382.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int t;
    cin >> t;
    
    while (t--) {
        int n;
        cin >> n;
        
        vector<int> arr(n + 1);
        vector<bool> visitado(n + 1, false);
        
        for (int i = 1; i <= n; i++) {
            cin >> arr[i];
        }
        
        int ciclos = 0;
        
        for (int i = 1; i <= n; i++) {
            if (!visitado[i]) {
                ciclos++;
                int j = i;
                while (!visitado[j]) {
                    visitado[j] = true;
                    j = arr[j];
                }
            }
        }
        
        cout << n - ciclos << '\n';
    }
    
    return 0;
}

File: uri/estructuras/elementary_my_dear_watson/1382.rs
use std::io::{self, BufRead, Write, BufWriter};

fn main() {
    let stdin = io::stdin();
    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());
    let mut lines = stdin.lock().lines();
    
    let t: i32 = lines.next().unwrap().unwrap().trim().parse().unwrap();
    
    for _ in 0..t {
        let n: usize = lines.next().unwrap().unwrap().trim().parse().unwrap();
        
        let arr: Vec<usize> = lines.next().unwrap().unwrap()
            .split_whitespace()
            .map(|x| x.parse().unwrap())
            .collect();
        
        let mut visitado = vec![false; n + 1];
        let mut ciclos = 0;
        
        for i in 1..=n {
            if !visitado[i] {
                ciclos += 1;
                let mut j = i;
                while !visitado[j] {
                    visitado[j] = true;
                    j = arr[j - 1];
                }
            }
        }
        
        writeln!(out, "{}", n - ciclos).unwrap();
    }
}
